operators/src/processing/expression.rs

```diff
@@ -0,0 +1,361 @@
+use crate::engine::{
+    InitializedOperator, InitializedOperatorImpl, InitializedRasterOperator, Operator,
+    QueryContext, QueryProcessor, QueryRectangle, RasterOperator, RasterQueryProcessor,
+    RasterResultDescriptor, TypedRasterQueryProcessor,
+};
+use crate::opencl::cl_program::{CLProgram, CompiledCLProgram, IterationType, RasterArgument};
+use crate::util::Result;
+use crate::{call_bi_generic_processor, call_generic_raster_processor};
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use geoengine_datatypes::raster::{Pixel, Raster, Raster2D, RasterDataType, RasterTile2D};
+use geoengine_datatypes::spatial_reference::SpatialReferenceOption;
+use serde::{Deserialize, Serialize};
+use snafu::ensure;
+use std::collections::HashSet;
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct ExpressionParams {
+    pub expression: String,
+    pub output_type: RasterDataType,
+}
+
+// TODO: custom type or simple string?
+struct SafeExpression {
+    expression: String,
+}
+
+pub type Expression = Operator<ExpressionParams>;
+
+impl Expression {
+    // TODO: also check this when creating original operator parameters
+    fn is_allowed_expression(expression: &str) -> bool {
+        // TODO: perform actual syntax checking
+        let disallowed_chars: HashSet<_> = ";[]{}#%\"\'\\".chars().collect();
+        let disallowed_strs: HashSet<String> =
+            vec!["/*".into(), "//".into(), "*/".into(), "return".into()]
+                .into_iter()
+                .collect();
+
+        expression.chars().all(|c| !disallowed_chars.contains(&c))
+            && disallowed_strs.iter().all(|s| !expression.contains(s))
+    }
+}
+
+#[typetag::serde]
+impl RasterOperator for Expression {
+    fn initialize(
+        self: Box<Self>,
+        context: crate::engine::ExecutionContext,
+    ) -> Result<Box<InitializedRasterOperator>> {
+        ensure!(
+            Self::is_allowed_expression(&self.params.expression),
+            crate::error::InvalidExpression
+        );
+
+        InitializedOperatorImpl::create(
+            self.params,
+            context,
+            |_, _, _, _| Ok(()),
+            |params, _, _, _, _| {
+                Ok(RasterResultDescriptor {
+                    data_type: params.output_type,
+                    spatial_reference: SpatialReferenceOption::None, // TODO
+                })
+            },
+            self.raster_sources,
+            vec![], // TODO: is it ok to omit them?
```

Comment:
I guess we should check that the input vector sources are empty.

Reply:
91ef24f5b869f75563586b32b932afa0a27fce0f

Reply:
You did not change this part in this commit.

Reply:
7ed0dcf6d9b9b5f6006b1f81b6d0fb44bb7a81c4